tests/lms/services/group_info_upsert_test.py

```diff
@@ -0,0 +1,100 @@
+from unittest import mock
+
+import pytest
+
+from lms.models import ApplicationInstance, GroupInfo
+from lms.services.group_info_upsert import GroupInfoUpsert
+
+
+class TestGroupInfoUpsert:
+    def test_it_adds_a_new_GroupInfo_if_none_exists(
+        self, db_session, group_info_upsert
+    ):
+        group_info_upsert(
+            "TEST_AUTHORITY_PROVIDED_ID",
+            "TEST_CONSUMER_KEY",
+            context_title="TEST_TITLE",
+            context_label="TEST_LABEL",
+        )
+
+        group_info = (
+            db_session.query(GroupInfo)
+            .filter_by(authority_provided_id="TEST_AUTHORITY_PROVIDED_ID")
+            .one()
+        )
+        assert group_info.consumer_key == "TEST_CONSUMER_KEY"
+        assert group_info.context_title == "TEST_TITLE"
+        assert group_info.context_label == "TEST_LABEL"
+
+    def test_it_updates_an_existing_GroupInfo_if_one_already_exists(
+        self, db_session, group_info_upsert
+    ):
+        db_session.add(
+            GroupInfo(
+                authority_provided_id="TEST_AUTHORITY_PROVIDED_ID",
+                consumer_key="TEST_CONSUMER_KEY",
+                context_title="ORIGINAL_TITLE",
+                context_label="ORIGINAL_LABEL",
+                tool_consumer_instance_name="ORIGINAL_NAME",
+            )
+        )
+
+        group_info_upsert(
+            "TEST_AUTHORITY_PROVIDED_ID",
+            "TEST_CONSUMER_KEY",
+            context_title="NEW_TITLE",
+            context_label="ORIGINAL_LABEL",
+        )
+
+        group_info = (
+            db_session.query(GroupInfo)
+            .filter_by(authority_provided_id="TEST_AUTHORITY_PROVIDED_ID")
+            .one()
+        )
+        assert group_info.consumer_key == "TEST_CONSUMER_KEY"
+        assert group_info.context_title == "NEW_TITLE"
+        assert group_info.context_label == "ORIGINAL_LABEL"
+        assert group_info.tool_consumer_instance_name == "ORIGINAL_NAME"
+
+    @pytest.fixture
+    def group_info_upsert(self, pyramid_config, pyramid_request):
+        """The group_info_upsert service that is being tested."""
+        return GroupInfoUpsert(mock.sentinel.context, pyramid_request)
+
+    @pytest.fixture
+    def application_instance(self, pyramid_request):
+        """The application instance that the test group infos belong to."""
+        application_instance = ApplicationInstance(
+            consumer_key="TEST_CONSUMER_KEY",
+            developer_key="TEST_DEVELOPER_KEY",
+            lms_url="TEST_LMS_URL",
+            shared_secret="TEST_SHARED_SECRET",
+            requesters_email="TEST_EMAIL",
+        )
+        pyramid_request.db.add(application_instance)
+        return application_instance
+
+    @pytest.fixture(autouse=True)
+    def group_infos(self, application_instance, pyramid_request):
+        """Add some "noise" group infos."""
```

Comment:
This is a super nice idea. Stops your tests from passing because you just retrieve *anything* that was lying around.